kafka_topic_enforcer/src/main/java/com/tesla/data/topic/enforcer/Main.java

```diff
@@ -14,17 +15,24 @@
 
 import java.util.HashMap;
 import java.util.Map;
+import java.util.ServiceLoader;
 
 /** The app. */
 public class Main {
 
   private static class TopicEnforceCommand extends EnforceCommand<ConfiguredTopic> {
     @Override
-    protected Enforcer<ConfiguredTopic> initEnforcer(AdminClient adminClient) {
-      return new TopicEnforcer(
-          new TopicServiceImpl(adminClient, dryrun),
-          configuredEntities(ConfiguredTopic.class, "topics", "topicsFile"),
-          !unsafemode);
+    protected Enforcer<ConfiguredTopic> initEnforcer(ServiceLoader<ClientProvider> clientLoader) {
+      // Iterate through all available client providers, and pick the one for kafka AdminClient
```

Comment:
instead, it can just specify in the constructor that it wanted the Kafka type admin client provider

Reply:
I think the issue here is that initEnforcer is called from EnforceCommand (not from the Acl/Topic EnforceCommand impls, which know what kind of client it needs), and so you can't just pass in which type of client it needs like that.

I think a reasonable workaround can be to just have initEnforcer take no arguments, and let the individual enforcer implementations deal with creating the clients that they need. This also drops the need for a provider-like interface, as the enforcer impls know exactly which clients they need, and can create them on the spot